packages/next/src/client/components/react-dev-overlay/ui/components/errors/error-overlay-toolbar/restart-server-button.tsx

```diff
@@ -0,0 +1,82 @@
+import { useState, useEffect } from 'react'
+import { RefreshClockWise } from '../../../icons/refresh-clock-wise'
+
+/**
+ * When the user reloads on a specific error and that error persists, we show
+ * the restart server button as an option. This is because some errors are
+ * recoverable by restarting the server and rebuilding the app.
+ *
+ * When Turbopack persistent cache is enabled, it will also clear the bundler
+ * cache. This improves DX by replacing the need to run `rm -rf .next` manually.
+ */
+export function RestartServerButton({ error }: { error: Error }) {
+  const [showButton, setShowButton] = useState(false)
+
+  useEffect(() => {
+    const ERROR_KEY = `error-overlay:${window.location.pathname}:${error.message}`
+
+    if (sessionStorage.getItem(ERROR_KEY) === '1') {
+      setShowButton(true)
+    } else {
+      setShowButton(false)
+    }
+
+    // When the user tries to reload, set the error key to the session storage.
+    const handleBeforeUnload = () => {
+      sessionStorage.setItem(ERROR_KEY, '1')
+    }
+    window.addEventListener('beforeunload', handleBeforeUnload)
+    return () => {
+      window.removeEventListener('beforeunload', handleBeforeUnload)
+    }
+  }, [error.message])
+
+  if (!showButton) {
+    return null
+  }
+
+  function handleClick() {
+    let endpoint = '/__nextjs_restart_dev'
+
+    if (process.env.__NEXT_TURBOPACK_PERSISTENT_CACHE) {
+      endpoint = '/__nextjs_restart_dev?invalidatePersistentCache'
+    }
+
+    fetch(endpoint, {
+      method: 'POST',
+    }).then(() => {
+      // TODO: poll server status and reload when the server is back up.
+      // https://github.com/vercel/next.js/pull/80005
```

Comment:
When we do, let's convert this to a Client Action so that we have a transition indicator. Need to wait for DevTools isolation to land because that needs React 19.